// @indigo-labs/dexter/build/dex/api/sundaeswap-api.js
import { BaseApi } from './base-api';
import { Asset } from '../models/asset';
import { LiquidityPool } from '../models/liquidity-pool';
import axios from 'axios';
import { SundaeSwap } from '../sundaeswap';
export class SundaeSwapApi extends BaseApi {
    constructor(dex, requestConfig) {
        super();
        this.dex = dex;
        this.api = axios.create({
            timeout: requestConfig.timeout,
            baseURL: `${requestConfig.proxyUrl}https://stats.sundaeswap.finance/graphql`,
            headers: {
                'Content-Type': 'application/json',
            }
        });
    }
    liquidityPools(assetA, assetB) {
        const maxPerPage = 100;
        const assetAId = (assetA === 'lovelace')
            ? ''
            : assetA.id('.');
        let assetBId = (assetB && assetB !== 'lovelace')
            ? assetB.id('.')
            : '';
        const getPaginatedResponse = (page) => {
            return this.api.post('', {
                operationName: 'getPoolsByAssetIds',
                query: `
                    query getPoolsByAssetIds($assetIds: [String!]!, $pageSize: Int, $page: Int) {
                        pools(assetIds: $assetIds, pageSize: $pageSize, page: $page) {
                            ...PoolFragment
                        }
                    }
                    fragment PoolFragment on Pool {
                        assetA {
                            ...AssetFragment
                        }
                        assetB {
                            ...AssetFragment
                        }
                        assetLP {
                            ...AssetFragment
                        }
                        name
                        fee
                        quantityA
                        quantityB
                        quantityLP
                        ident
                        assetID
                    }
                    fragment AssetFragment on Asset {
                        assetId
                        decimals
                    }
                `,
                variables: {
                    page: page,
                    pageSize: maxPerPage,
                    assetIds: [assetBId !== '' ? assetBId : assetAId],
                },
            }).then((response) => {
                const pools = response.data.data.pools;
                const liquidityPools = pools.map((pool) => {
                    let liquidityPool = new LiquidityPool(SundaeSwap.identifier, pool.assetA.assetId
                        ? Asset.fromId(pool.assetA.assetId, pool.assetA.decimals)
                        : 'lovelace', pool.assetB.assetId
                        ? Asset.fromId(pool.assetB.assetId, pool.assetB.decimals)
                        : 'lovelace', BigInt(pool.quantityA), BigInt(pool.quantityB), this.dex.poolAddress, this.dex.orderAddress, this.dex.orderAddress);
                    liquidityPool.identifier = pool.ident;
                    liquidityPool.lpToken = Asset.fromId(pool.assetLP.assetId);
                    liquidityPool.poolFeePercent = Number(pool.fee);
                    return liquidityPool;
                });
                if (pools.length < maxPerPage) {
                    return liquidityPools;
                }
                return getPaginatedResponse(page + 1).then((nextPagePools) => {
                    return liquidityPools.concat(nextPagePools);
                });
            });
        };
        return getPaginatedResponse(0);
    }
}
